### ISSUE TITLE
Question about data augmentations

### ISSUE THREAD

**Issue by cactusgame** [tanguymagne/UVDoc#12]
Hi, thanks for your great work!

I'm confused by the fact that there are three data augmentation options: "rotate," "flip," and "perspective." The "perspective" option seems to not work. 

I find the `Dataset` code below, the `transform` is done with `grid2d` and `image` together, but without `grid3d`. If I open the "perspective" option, should I do some work on "grid3d"? 

```
        # Geometric Augmentations
        with warnings.catch_warnings():
            warnings.filterwarnings("ignore")
            transformed = self.geometric_transform(  # 透视，旋转等更有用的增强
                image=img_RGB_,
                keypoints=grid2D_.transpose(1, 2, 0).reshape(-1, 2),  # default: 89*61
            )
            img_RGB_ = transformed["image"]
            grid2D_ = np.array(transformed["keypoints"]).reshape(*self.original_grid_size, 2).transpose(2, 0,
                                                                                                        1)  # 89*61
            flipped = False
            for x in transformed["replay"]["transforms"]:
                if "SafeHorizontalFlip" in x["__class_fullname__"]:
                    flipped = x["applied"]
            if flipped:
                grid3D[1] = 1 - grid3D[1]
                grid3D = torch.flip(grid3D, dims=(2,))
```

Btw, I've trained a model with `perspective` for 30 epochs with doc3d and UVDoc dataset, and the results are quite strange from the third epoch. There are some unwarp images predicted by the model trained by `perspective`. 

![image](https://github.com/user-attachments/assets/a6b8cd33-a8d9-4e02-bd1a-4c8c9127b42a)

![image](https://github.com/user-attachments/assets/35b66234-6023-4d38-90fb-9f0f70685a7d)

**Comment by ruanmk** [comment]
I think turning off the grid3d loss and its relative data preprocessing should work better. we may also use depth map as another constrain(from the camera space). It seems like the author treats all document paper objects in  its own object space, regardless of the camera or viewpoint, but its hard to get such kind of data in real world.

**Comment by cactusgame** [comment]
Thanks for your reply.

I find that the training data used in the repo makes the 4 edges of all documents visible. If I crop the training data and the backward map, the model can learn to the "perspective" data. But the prediction often has a white border like the image below, 

![image](https://github.com/user-attachments/assets/a9fb5975-3c82-4acc-ab23-325506deff9a)


> I think turning off the grid3d loss and its relative data preprocessing should work better. we may also use depth map as another constrain(from the camera space). It seems like the author treats all document paper objects in its own object space, regardless of the camera or viewpoint, but its hard to get such kind of data in real world.

**Comment by tanguymagne** [comment]
> Hi, thanks for your great work!
> 
> I'm confused by the fact that there are three data augmentation options: "rotate," "flip," and "perspective." The "perspective" option seems to not work.
> 
> I find the `Dataset` code below, the `transform` is done with `grid2d` and `image` together, but without `grid3d`. If I open the "perspective" option, should I do some work on "grid3d"?

In theory, you shouldn't have to do anything to the "grid3d" even with _perspective_ transform. This is because, as mentioned by @ruanmk we indeed treat all document paper objects in their own object space.

> Btw, I've trained a model with `perspective` for 30 epochs with doc3d and UVDoc dataset, and the results are quite strange from the third epoch. There are some unwarp images predicted by the model trained by `perspective`.

The results are indeed a bit strange. We also observed that adding _perspective_ augmentations didn't help the network, which is why we trained our final network without this augmentation (as you can see in [train.py](https://github.com/tanguymagne/UVDoc/blob/main/train.py#L354) by default we only use _rotate_ augmentation). 

However, a better representation of the 3D grid might exist and may require to be modified under perspective augmentation.

I hope this helps a little bit, I don't have a definitive answer to your questions, since those are research questions and you would have to experiment a bit yourself.
